Tests/kaas/k8s-default-storage-class/k8s-default-storage-class-check.py

```diff
@@ -180,26 +182,112 @@ def check_default_persistentvolumeclaim_readwriteonce(k8s_api_instance, storage_
+            logger.info(f"L{inspect.currentframe().f_lineno} {exception_message}")
+            self.return_message = f"{exception_message}"
+            self.return_code = 1
+
+        logger.debug("Checking Environment for Leftovers")
+        try:
+            pod_list = self.k8s_core_api.list_namespaced_pod(namespace=self.namespace)
+            for pod in pod_list.items:
+                if pod.metadata.name == self.pod_name:
+                    logger.debug(
+                        f"POD '{self.pod_name}' exists in namespace '{self.namespace}'"
+                    )
+                    return True
+            pvc_list = self.k8s_core_api.list_namespaced_persistent_volume_claim(
+                namespace=self.namespace
+            )
+            for pvc in pvc_list.items:
+                if pvc.metadata.name == self.pvc_name:
+                    logger.debug(
+                        f"PVC '{self.pvc_name}' exists in namespace '{self.namespace}'"
+                    )
+                    return True
+            return False
+        except ApiException as e:
+            logger.debug(f"Error preparing Environment: {e}")
+            return False
+
+    def clean(self):
+        api_response = None
+        try:
+            logger.debug(f"delete pod:{self.pod_name}")
+            api_response = self.k8s_core_api.delete_namespaced_pod(
+                self.pod_name, self.namespace
+            )
+        except:
+            logger.debug(f"The pod {self.pod_name} couldn't be deleted.", exc_info=True)
+        try:
+            logger.debug(f"delete pvc:{self.pvc_name}")
+            api_response = self.k8s_core_api.delete_namespaced_persistent_volume_claim(
+                self.pvc_name, self.namespace
+            )
+        except:
+            logger.debug(f"The PVC {self.pvc_name} couldn't be deleted.", exc_info=True)
+        return api_response
+
+    def __enter__(self):
+        retries = 0
+        while retries <= 2:
+            if self.prepare():
+                self.clean()
+                logger.debug(
+                    f"Deleting Leftovers in namespace {self.namespace} from previous test runs"
+                )
+                time.sleep(2)
+            else:
+                logger.debug(f"Entering the context {self.k8s_core_api}")
+                return self
+            retries += 1
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.clean()
+        if self.return_code == 0:
+            self.return_message = "all tests passed"
+        if isinstance(exc_value, SCSTestException):
+            self.return_message = exc_value.args[0]
+            self.return_code = exc_value.return_code
+            print(f"SCSTestException occurred with return_code: {self.return_code}")
+        else:
+            # No specific exception, handle normally
+            print(f"Exiting the context with return_code: {self.return_code}")
```

Comment:
This seems really convoluted. The exceptions are being caught in the main function anyway. And why should the test environment carry the burden of storing the return code? That violates the single-responsibility principle (aka separation of concerns).

Reply:
because in this case I assumed that these exceptions are only intercepted afterwards. i'll check this again.

Reply:
the return code is kept there to feed it into the sonobuoy result, which is created in the exit, to make sure it gets created, after any error whatsoever. This is actually the reason why i created this context management to avoid having the same exception (SCSTestException) raised in 5 different places.